Replicator/Replicator.cc

```diff
@@ -1148,35 +1123,48 @@ namespace litecore { namespace repl {
             request->respondWithError({"BLIP"_sl, 400, "Invalid getCollections message: mismatched checkpoint_ids and collections"_sl});
             return;
         }
-        
-        if (_sessionCollections.size() > 0) {
-            request->respondWithError({"BLIP"_sl, 400, "getCollections is already settled for the replicator"_sl});
-            return;
+
+        // Convert to collection specs
+        vector<C4CollectionSpec> collSpecs;
+        collSpecs.reserve(collections.count());
+        for (fleece::Array::iterator i(collections); i; ++i) {
+            if (!i.value().asString()) {
+                request->respondWithError({"BLIP"_sl, 400, "Invalid getCollections message: empty collection path"_sl});
+                return;
+            }
+            collSpecs.emplace_back(Options::collectionPathToSpec(i.value().asString()));
         }
         
-        unordered_map<C4Database::CollectionSpec, C4Collection*> collectionMap;
-        for (int i = 0; i < _subRepls.size(); i++) {
-            C4Collection* coll = collection(i);
-            collectionMap.insert({coll->getSpec(), coll});
+        // check duplicates
+        std::unordered_set<C4CollectionSpec> specSet;
+        for (size_t i = 0; i < collSpecs.size(); ++i) {
+            auto [it, b] = specSet.insert(collSpecs[i]);
+            if (!b) {
+                request->respondWithError({"BLIP"_sl, 400, "Invalid getCollections message: duplicate collection path"_sl});
+                return;
+            }
         }
-        
+
+        // Rearrange _options->collectionOpts according to collSpecs. If i-th CollectionSpec
+        // is not found, put an empty collectionOptions at i-th position.
+        _options->rearrangeCollections(collSpecs);
```

Comment:
I'm not a big fan of mutating the input configuration. This will work if _options is a copied config for the replicator. It's a pointer so I'm not sure.

Reply:
(meant to insert.a link here but ...)
        // RearrangeCollections() is called only by the passive replicator. For the passive replicator, we presume
        // that the order of the collection properties is not important. So, we take it as legit to permutate
        // it in const method. (The Replicator holds a const Options object.) It is supposed to be called as it
        // starts to interact with the active replicator.
        // "collections" is a list of CollectionSpecs that the active replicatore proposes to replicate, and the
        // order will be used as index to refer to respective collections.
        // Post-conditions: collectionOpts[i] and collections[i] share the same CollectionSpec if
        //                    collections[i] is found in collectionOpts;
        //                  Otherwise, an empty collectionOptions is inserted in collectionOpts[i].
        // By empty collectionOptions we mean the collection path is a null slice.

Reply:
Let's discuss it in the meeting